#' Plot genomes, features and synteny maps
#'
#' @param data genome layout generated by `layout_genomes()`
#' @param theme choose a gggenomes default theme, NULL to omit.
#' @param scale choose a gggenomes dafault y-scale, NULL to omit.
#' @inheritParams ggplot2::ggplot
#' @param ... passed through to contig layout
#' @importFrom ggplot2 ggplot
#' @export
#' @return ggplot object
#' @param ... layout parameter
gggenomes <- function(contigs, genes = NULL, links = NULL, ...,
    scale = c("lab", "numlab", NULL), theme = c("clean", NULL),
    mapping = aes(), environment = parent.frame()){

  layout <- as_genomes(contigs=contigs, genes=genes, links=links, ...)
  
  p <- ggplot(data = layout, mapping = mapping, environment = environment)
  class(p) <- c('gggenomes', class(p))
  
  scale_name <- scale[[1]] %||% match.arg(scale[[1]], c("lab", "numlab"))
  if(!is.null(scale_name)){ # add gggenomes scale
    scale_args <- if(is.list(scale) && length(scale) >1) scale[-1] else list()
    p$data[["ggargs_"]]$scale_args <- scale_args # store for recomputation on re-layout
    p$data[["ggargs_"]]$scale_name <- scale_name
    scale_args$data <- layout
    p <- p + do.call(paste0("scale_gggenomes_", scale_name), scale_args)
  }
  
  theme_name <- theme[[1]] %||% match.arg(theme[[1]], c("clean"))
  if(!is.null(theme_name)){ # add theme
    theme_args <- if(is.list(theme) && length(theme) >1) theme[-1] else list()
    p <- p + do.call(paste0("theme_gggenomes_", theme), theme_args)
  }
  
  p
}

#' @export
layout.gggenomes <- function(x){
  x$data <- layout(x$data)
  scale_name <- x$data[["ggargs_"]]$scale_name
  scale_args <- x$data[["ggargs_"]]$scale_args
  scale_args$data <- x$data
  x <- x + do.call(paste0("scale_gggenomes_", scale_name), scale_args)
  x
}

#' ggplot.default tries to `fortify(data)` and we don't want that
#'
#' @export
ggplot.tbl_genome <- function(data, mapping = aes(), ...,
                               environment = parent.frame()) {
  if (!missing(mapping) && !inherits(mapping, "uneval")) {
    stop("Mapping should be created with `aes() or `aes_()`.", call. = FALSE)
  }

  p <- structure(list(
    data = data,
    layers = list(),
    scales = ggplot2:::scales_list(),
    mapping = mapping,
    theme = list(),
    coordinates = coord_cartesian(default = TRUE),
    facet = facet_null(),
    plot_env = environment
  ), class = c("gg", "ggplot"))

  p$labels <- ggplot2:::make_labels(mapping)

  ggplot2:::set_last_plot(p)
  p
}

#' gggenomes default theme
#' @importFrom ggplot2 theme_bw
#' @importFrom ggplot2 theme
#' @inheritParams ggplot2::theme_bw
#' @export
theme_gggenomes_clean <- function(base_size = 24, base_family = "", base_line_size = base_size/22, base_rect_size = base_size/22){
  theme_bw(
    base_size = base_size, base_family = base_family, 
    base_line_size = base_line_size, base_rect_size = base_rect_size
  ) + theme(
    panel.border = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    panel.background = element_rect(fill = "white"),
    axis.ticks.x=element_blank(),
    axis.title.y=element_blank(),
    axis.ticks.y=element_blank())
}
#' y scale with genome labels
#'
#' @name scale_gggenomes
NULL

#' @rdname scale_gggenomes
#' @import ggplot2
#' @export
scale_gggenomes_lab <- function(data, limits = NULL, ...){
  dd <- unique(select(data$contigs, .gix, genome_id))
  breaks <- dd$.gix
  labels <- dd$genome_id

  limits <- limits %||% c(max(breaks) + .5, min(breaks) -.5)
  scale_y_continuous("", limits = limits, breaks = breaks,
    labels = labels, trans = scales::reverse_trans(), ...)
}
#' @rdname scale_gggenomes
#' @export
scale_gggenomes_numlab <- function(data, limits = NULL, ...){
    breaks <- unique(expose(data)$.gix)
    labels <- paste("[", breaks, "] ", unique(expose(data)$genome_id), sep="")
    limits <- limits %||% c(max(breaks) + .5, min(breaks) -.5)
    scale_y_continuous("", limits = limits, breaks = breaks,
        labels = labels, trans = scales::reverse_trans(), ...)
}
